Added Virtuoso support for d1lod service

rushirajnenuji opened the pull request.

rushirajnenuji committed:
Ignoring PyCharm and VirtualEnv files

rushirajnenuji committed:
Adding the Virtuoso docker image to the docker-compose

rushirajnenuji committed:
[init] Graph Util module to manage d1lod contents in a graph.

Adding a utility module for the d1lod service to read and write d1lod contents to the Virtuoso graphs.

rushirajnenuji committed:
Added GraphManager.py

This is a graph manager for the d1lod service to manage the Virtuoso graphs.

rushirajnenuji committed:
Updating requirements.txt

rushirajnenuji committed:
Re-factoring Store and Repository module

Re-factoring Store and Repository module to make it work with Virtuoso database

rushirajnenuji committed:
Using the Virtuoso Store and Repository module

Refactoring the Interface to make use of the Virtuoso Store and Repository module

rushirajnenuji committed:
Adding sample turtle dataset to test Virtuoso

rushirajnenuji committed:
Refactoring Docker-Compose - Adding Virtuoso

rushirajnenuji committed:
Setting Virtuoso end-point for Repository and Store instances

rushirajnenuji committed:
Checking for blank nodes while inserting triples

rushirajnenuji committed:
Updating the syntax of delete query

Delete query had wrong syntax and Virtuoso was giving Error 37000 because of wrong syntax. This commit fixes the query syntax.

rushirajnenuji committed:
Removing the log statement which was meant only for testing.

Removing the log statement which was meant only for testing.

rushirajnenuji committed:
Fetching Sci Metadata from SOLR

amoeba commented on d1lod/d1lod/jobs.py:
```diff
@@ -51,9 +51,9 @@
-SESAME_HOST = os.getenv('WEB_1_PORT_8080_TCP_ADDR', 'localhost')
-SESAME_PORT = os.getenv('WEB_1_PORT_8080_TCP_PORT', '8080')
-SESAME_REPOSITORY = 'd1lod'
+VIRTUOSO_HOST = "virtuoso"
```
My original implementation used env vars which allowed some flexibility that I liked. Looking at the code, probably incorrectly. I don't remember why I wrote `WEB_1_PORT_8080_TCP_ADDR` which looks hacky and messy. Since Docker seamlessly manages the  `virtuoso` hostname for us, this change is much better!

It'd be nice, though, if this and the lines of code below it, failed over better for local development. If the env vars aren't set, I'd like this script to fall back to `localhost:8890` so that I can run individual components outside of Docker against a locally installed Virtuoso (or other SPARQL endpoint) if needed.

Since we're abstracting this to any SPARQL endpoint, it seems like HOST/PORT/REPOSITORY could be all combined into just an env var like "SPARQL_URI" with a value set in the `docker-compose.yml` like `SPARQL_URI=http://virtuoso:8890/sparql`.

What do you think?

amoeba commented on d1lod/d1lod/jobs.py:
```diff
@@ -213,11 +213,13 @@ def calculate_stats():
     Interface(r)  # Adds namespaces we need to repo
 
-    logging.info("[%s] repository.size=%d", JOB_NAME, r.size())
+    # have no feature to determine the repository size
```
Could this either get implemented and put back in or just removed since we're using version control and don't really needed commented-out code.

rushirajnenuji replied:
Resolved by implementing the `size()` function

amoeba commented on d1lod/d1lod/jobs.py:
```diff
@@ -362,10 +365,10 @@ def export_graph():
+    r = Repository(host=VIRTUOSO_HOST, port=VIRTUOSO_PORT, name=VIRTUOSO_REPOSITORY, ns=NAMESPACES)
 
-    logging.info("[%s] Exporting graph of size %d.", JOB_NAME, r.size())
+    # logging.info("[%s] Exporting graph of size %d.", JOB_NAME, r.size())
```
Could this either get implemented and put back in or just removed since we're using version control and don't really needed commented-out code.

rushirajnenuji replied:
Resolved by implementing the `size()` function

amoeba commented on d1lod/d1lod/sesame/interface.py:
```diff
@@ -185,13 +175,20 @@ def insertModel(self):
+        blank_node = False
+        for s in self.model:
+            blank_node = self.tripleHasBlankNode(s.subject, s.predicate, s.object)
+            if(blank_node == True):
```
Unnecessary parens around conditional:

```
if(blank_node == True):
```

to

```
if blank_node == True:
```

rushirajnenuji replied:
Made conditions Python2 compliant.
Commit: https://github.com/ec-geolink/d1lod/pull/48/commits/be01976d391a3f32b6e6b63c6f6dfde7d57988cd

amoeba commented on d1lod/d1lod/sesame/interface.py:
```diff
@@ -185,13 +175,20 @@ def insertModel(self):
         logging.info('Inserting model of size %d.', self.model.size())
 
-        self.repository.update(sparql_query)
+        self.repository.insert_data(repository_name=self.repository.name,payload=sparql_data, blank_node = blank_node)
```
Extra space around arg

```
blank_node = blank_node
```

to

```
blank_node=blank_node
```

rushirajnenuji replied:
Made conditions Python2 compliant.
Commit: https://github.com/ec-geolink/d1lod/pull/48/commits/be01976d391a3f32b6e6b63c6f6dfde7d57988cd

amoeba commented on d1lod/d1lod/sesame/interface.py:
```diff
@@ -573,12 +577,15 @@ def deleteDatasetTriples(self, identifier):
+            }
         }
-        """ % (dataset, has_identifier)
+        """ % (self.repository.name, dataset, has_identifier)
```
I see lots of explicitly passing in the repository (graph) name to functions handling the execution of SPARQL queries. Since you're passing `self.repository_name` to another method on `self`, which already knows which `repository` it's pointing at, whichever method you're calling could just automatically know which repository you mean. Do these method signatures need the flexibility to insert triples into another graph?

Calls like `self.repository.insert_data(repository_name=self.repository.name` seem like we're missing an opportunity to be more concise like `self.repository.insert_data(dataset, something)`

rushirajnenuji replied:
Removed functionality to be able to insert another graph into the DB.
Commit: https://github.com/ec-geolink/d1lod/pull/48/commits/a6e0e1f40f2d88ab3f8211431534a8bf1174aaee

amoeba commented on d1lod/d1lod/sesame/store.py:
```diff
@@ -1,150 +1,334 @@
 
-        return r.text
 
+    def create_repository(self, repository_name, silent=False):
```
Since we're no longer targeting Sesame, can we move away from the name Repository which doesn't have meaning outside of Sesame? I think the new equivalent is "Graph" or "Context" and I think Graph is most appropriate here.

rushirajnenuji replied:
Refactored Graph and Interface modules and moved them to the default package.
Commit: https://github.com/ec-geolink/d1lod/pull/48/commits/83ea088f4d336fb606422dc6e0117c7f108ed410

amoeba reviewed (changes requested):
Looking great, thanks. I think the bulk of my comments were about the separation between what a Store/Repository/Interface is and what each does. Since I introduced these classes to model Sesame and we're no longer using Sesame, I think these might be the wrong classes to model all of the logic going on here.

Do you think all three of these classes are still needed, or do you think we could add more or remove some, perhaps giving them better names, might be an improvement?

I didn't test running the service, or run the tests as I figured you were still making some changes there. Lemme know if/when you want me to try those out.

rushirajnenuji committed:
Refactoring package name to Virtuoso throughout the code base

Refactored package name to Virtuoso and tested the d1lod service. The service seems to be working as expected after commiting this change.

rushirajnenuji committed:
Adding a function to get list of existing reporitories

Adding a function to get list of existing reporitories

rushirajnenuji committed:
Refactoring code to create independent Graph and Interface modules

- Refactored Virtuoso package and made Graph and Interface modules independent.
- Changes suggested from previous PR feedback

rushirajnenuji committed:
Updated Tests for the d1lod service

Updated Tests for the d1lod service

rushirajnenuji committed:
Remove functionality to insert triples to another graph

Remove functionality to insert triples to another graph

rushirajnenuji committed:
Updating test cases for the Graph Module

rushirajnenuji committed:
Updating the test suite

Updating the test suite

rushirajnenuji committed:
Making conditions Python2 compliant

Making conditions Python2 compliant

rushirajnenuji committed:
Removing relative paths from metadata test

Removing relative paths from metadata test

amoeba commented on d1lod/README.md:
```diff
@@ -8,13 +8,14 @@ This directory contains the Python package that supports the LOD service.
 ### Pre-requisites:
 
-- Sesame w/ GraphDB must be running on 'http://localhost:8000/openrdf-workbench'
+- Virtuoso Store must be running on 'http://localhost:8000/virtuoso/conductor'
```
I think with our setup, this should be 8890. I can change this after merge.

amoeba commented on d1lod/scripts/benchmark.py:
```diff
@@ -2,8 +2,8 @@
 
-s = sesamestore.SesameStore("localhost", 8080)
-r = sesamestore.SesameRepository(s, "test")
+s = sesamestore.SesameStore("virtuoso", 8890)
```
This seems like it'd no longer be relevant since we stripped out Sesame. I'll take a look at this after merge.

amoeba commented:
Looks good @rushirajnenuji 	, thanks for all the hard work on the refactor!
